pkg/util/util.go

```diff
@@ -302,3 +305,24 @@ func WaitUntilTimeout(name string, fn func(), timeout time.Duration) {
 	case <-exited:
 	}
 }
+
+// WriteFileAtomic writes file to temp and atomically move when everything else succeeds.
+func WriteFileAtomic(filename string, data []byte, perm os.FileMode) error {
+	dir, name := path.Dir(filename), path.Base(filename)
+	f, err := os.CreateTemp(dir, name)
+	if err != nil {
+		return err
+	}
+	n, err := f.Write(data)
+	f.Close()
+	if err == nil && n < len(data) {
```

Comment:
Write must return a non-nil error if it returns n < len(p)
https://pkg.go.dev/io#Writer


this is an unreachable branch

Reply:
I push a followup to fix this picked method. PTAL https://github.com/pingcap/tidb-binlog/pull/1048/commits/b591b0a7f6d50eead9bd49de8f9ac9d627bb0c19